src/lsm/grid.zig

```diff
@@ -392,10 +479,10 @@ pub fn GridType(comptime Storage: type) type {
             assert(address > 0);
             assert(address == header.op);
             assert(grid.superblock.opened);
+            maybe(grid.cancelling);
```

Comment:
Hm, it feels odd that we allow write_repair while we are cancelling, for two reasons: 

* we are about to sync to a different checkpoint, so it's not clear if this write is useful
* we don't want to wait for all repairs to complete, so waiting for none to complete reduces the state space. 

Should we perhaps clear `read_faulty_queue` when we enter `grid.cancel()`, rather when we exit `cancel_join_callback`?

Reply:
Good point; there's no reason to allow grid repair during cancellation.

I am going to implement it by checking `grid.cancelling` in `Replica.on_block`, though.
The reason to not `read_faulty_queue` when we enter `grid.cancel()` is that it avoids some `grid.cancelling` special-casing in `grid.write_block_callback` -- after `grid.cancel()` is called, we just allow any writes that were already in progress to complete like normal. (Part of that involves repairs dequeuing themselves from the `read_faulty_queue`).